services/rpcfilters/logs_cache_test.go

```diff
@@ -0,0 +1,119 @@
+package rpcfilters
+
+import (
+	"testing"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestAggregateLogs(t *testing.T) {
+	logs := []types.Log{}
+	for i := 1; i <= 15; i++ {
+		logs = append(logs,
+			types.Log{BlockNumber: uint64(i), BlockHash: common.Hash{byte(i)}},
+			types.Log{BlockNumber: uint64(i), BlockHash: common.Hash{byte(i)}})
+	}
+	aggregated := aggregateLogs(logs, 10)
+	start := 15 - len(aggregated) + 1
+	for _, record := range aggregated {
+		assert.Equal(t, start, int(record.block)) // numbers are small
+		assert.Len(t, record.logs, 2)
+		start++
+	}
+}
+
+func TestAggregateLessThenFull(t *testing.T) {
+	logs := []types.Log{}
+	for i := 1; i <= 3; i++ {
+		logs = append(logs,
+			types.Log{BlockNumber: uint64(i), BlockHash: common.Hash{byte(i)}})
+	}
+	aggregated := aggregateLogs(logs, 10)
+	start := 3 - len(aggregated) + 1
+	for _, record := range aggregated {
+		assert.Equal(t, start, int(record.block)) // numbers are small
+		assert.Len(t, record.logs, 1)
+		start++
+	}
+}
+
+func TestMerge(t *testing.T) {
+	step1Logs := []types.Log{}
+	for i := 1; i <= 3; i++ {
```

Comment:
Just my 2c. `for` loop is cool because it's dynamic but from the reader perspective, it's a bit hard to figure out what we compare and expects in the asserts below. Maybe unfolding the loops would be helpful at least in this case, at we generate just 2 or 3 values.

Reply:
sure, would prefer unfolded values myself